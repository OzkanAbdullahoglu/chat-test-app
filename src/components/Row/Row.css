.list-end-message {
  height: 100px;
}
