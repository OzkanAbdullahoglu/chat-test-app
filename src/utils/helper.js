/* eslint-disable linebreak-style */
import React, { useState, useEffect } from 'react';
export const timeConverter = (timeStamp) => {
  if (!timeConverter.cache) {
    timeConverter.cache = {};
  }

  if (!timeConverter.cache[timeStamp]) {
    const timeStampMs = timeStamp * 1000;
    const dateObj = new Date(timeStampMs);
    const dayOfMonth = dateObj.getDate() < 10 ? `0${dateObj.getDate()}` : dateObj.getDate();
    const months = [
      'January', 'February', 'March', 'April', 'May', 'June',
      'July', 'August', 'September', 'October', 'November', 'December',
    ];
    const curMonth = months[dateObj.getMonth()];
    const curYear = dateObj.getFullYear();
    const curMinute = dateObj.getMinutes() < 10 ? `0${dateObj.getMinutes()}` : dateObj.getMinutes();
    const curHour = (dateObj.getHours() < 10 ? `0${dateObj.getHours()}` : dateObj.getHours());

    timeConverter.cache[timeStamp] = {
      date: `${curMonth} ${dayOfMonth},${curYear}`,
      time: `${curHour}:${curMinute}`,
    };
    return timeConverter.cache[timeStamp];
  }
  return timeConverter.cache[timeStamp];
};

export const isToday = (timestamp, date) =>{
  const today = Math.floor(Date.now() / 1000).toString();
  console.log(today);
  return timeConverter(today).date === timeConverter(timestamp).date ?
    'Today' : timeConverter(today).date === timeConverter(timestamp).date - 1 ?
    'Yesterday' : date;
};




function getCurrentViewDims() {
  const { innerWidth: width, innerHeight: height } = window;
  return {
    width,
    height,
  };
}

export default function useCurrentViewDims() {
  const [windowDimensions, setWindowDimensions] = useState(getCurrentViewDims());

  useEffect(() => {
    function handleResize() {
      setWindowDimensions(getCurrentViewDims());
    }

    window.addEventListener('scroll', handleResize);
    return () => window.removeEventListener('scroll', handleResize);
  }, []);

  return windowDimensions;
}

