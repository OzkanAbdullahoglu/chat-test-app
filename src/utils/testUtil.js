/* eslint-disable linebreak-style */
import { createStore, applyMiddleware } from 'redux';
import { middlewares } from '../store';
import rootReducer from '../reducers';

const chat = {
  requestedData: [
    {
      id: 1,
      direction: 'in',
      status: 'read',
      timestamp: '1577834102',
      text: 'Alice was beginning',
    },
  ],
  isPending: false,
  error: '',
  scrollDownEnable: true,
};

const message = {
  messagesEndRef: '',
};

const version = { number: '0.0.1' };

export const initialStates = { chat, message, version };

export const storeFactory = (initalState) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, initalState);
};
