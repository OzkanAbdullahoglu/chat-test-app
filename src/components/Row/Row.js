/* eslint-disable linebreak-style */
import React, { useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import PropTypes from 'prop-types';

import Message from '../Message/Message';
import '../Row/Row.css';
import Date from '../Date/Date';
import {
  messageActions,
  getRequestedData,
  getShowTimeStampBool,
} from '../../reducers';

const Row = ({
  index,
  style,
  showTimeStampBool,
  updatedInitialData,
  setLastMessageRef }) => {
  const items = updatedInitialData;
  const messagesEnd = useRef();
  useEffect(() => {
    if (index === items.length - 1) {
      setLastMessageRef(messagesEnd);
    }
  }, []
  );
  return (
    showTimeStampBool(items[index].id) ? (
      <div style={style} className={`list-item-container ${index}`} key={items[index].id}>
        <Date pack={items[index]} />
        <Message pack={items[index]} />
        {index === items.length - 1 ? (
          <div className="list-end-message list-item-container" ref={messagesEnd}></div>
        ) : null
        }
      </div>
    ) : (
      <div style={style} className={`list-item-container ${index}`} key={items[index].id}>
        <Message pack={items[index]} />
        {index === items.length - 1 ? (
          <div className="list-end-message list-item-container" ref={messagesEnd}></div>
        ) : null
        }
      </div>
    )
  );
};

Row.propTypes = {
  updatedInitialData: PropTypes.array,
  showTimeStampBool: PropTypes.func,
  setLastMessageRef: PropTypes.func,
  index: PropTypes.number,
  style: PropTypes.object,
};

const mapStateToProps = (store) => ({
  updatedInitialData: getRequestedData(store),
  showTimeStampBool: getShowTimeStampBool(store),
});

const withRedux = connect(
  mapStateToProps,
  { ...messageActions },
  null,
  { forwardRef: true },
);

export default compose(withRedux)(Row);
