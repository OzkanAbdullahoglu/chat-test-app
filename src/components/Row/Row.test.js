/* eslint-disable linebreak-style */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Row from './Row';
import { storeFactory } from '../../utils/testUtil';

describe('Row Component', () => {
  const store = storeFactory();
  const mockFunc = jest.fn();
  const props = {
    index: 4,
    showTimeStampBool: mockFunc,
    updatedInitialData: [
      {
        id: 1,
        direction: 'in',
        status: 'read',
        timestamp: '1577834102',
        text: 'Alice was beginning',
      },
      {
        id: 2,
        direction: 'out',
        status: 'received',
        timestamp: '1577834168',
        text: 'to get very tired',
      },
      {
        id: 3,
        direction: 'in',
        status: 'read',
        timestamp: '1577834198',
        text: 'of sitting by her',
      },
      {
        id: 4,
        direction: 'in',
        status: 'read',
        timestamp: '1577834323',
        text: 'sister on the',
      },
      {
        id: 5,
        direction: 'in',
        status: 'read',
        timestamp: '1577834479',
        text: 'bank, and of having nothing to',
      },
    ],
  };
  const row = shallow(<Row store={store} {...props} />)
    .dive()
    .dive();
  it('renders without crashing', () => {
    expect(toJson(row)).toMatchSnapshot();
  });
});
