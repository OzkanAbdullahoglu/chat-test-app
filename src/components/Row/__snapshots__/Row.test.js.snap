// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row Component renders without crashing 1`] = `
<Row
  index={4}
  setLastMessageRef={[Function]}
  showTimeStampBool={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  updatedInitialData={Array []}
/>
`;
